### ISSUE TITLE
TPC-C: original stock level procedure crashes

### ISSUE THREAD

**Issue by Bensk1** [hyrise/hyrise#2183]
The bachelor project students (thanks to Alexander and Petr) are trying to get PyTpcc to work over the network. For multiple warehouses the stock level procedure is failing. This can be reproduced in the console with:

```
generate_tpcc 5
SELECT COUNT(DISTINCT(OL_I_ID)) FROM ORDER_LINE, STOCK WHERE OL_W_ID = 1 AND OL_D_ID = 9 AND OL_O_ID < 3006 AND OL_O_ID >= 2986 AND S_W_ID = 1 AND S_I_ID = OL_I_ID AND S_QUANTITY < 10;

libc++abi.dylib: terminating with uncaught exception of type std::logic_error: src/lib/statistics/cardinality_estimator.cpp:987 Semi join should not increase cardinality
```

While looking at it, I saw that the Hyrise TPC-C Stock Level procedure is not implemented in the same way. PyTpcc is closer to the [specification](http://www.tpc.org/tpc_documents_current_versions/pdf/tpc-c_v5.11.0.pdf) (page 116). @mrks do you remember the reasoning behind this decision?

**Comment by mrks** [comment]
Regarding the CardinalityEstimator: Yes, I think that the assertion is too strict. If you print `output_table_statistics->row_count`, I'd expect it to be less than one. We should probably accept cardinalities `< 1`, too.

Regarding the Stock Level Procedure: I did not look at the example code. Doing this as a join seems sensible. No idea why I didnt'.

**Comment by Bensk1** [comment]
You are right, in this case `left_input_table_statistics.row_count * 1.01f` is `-661904`.

**Comment by mrks** [comment]
Hmm, I would have thought that it is close to zero. How can cardinalities be negative? That sounds like there is something wrong there.

**Comment by Bensk1** [comment]
> That sounds like there is something wrong there.

I agree.

> How can cardinalities be negative?

I have no idea. I believe that @aloeser and/or @Bouncner have seen negative cardinalities before but only slightly negative ones.

**Comment by aloeser** [comment]
>I have no idea. I believe that @aloeser and/or @Bouncner have seen negative cardinalities before but only slightly negative ones.
I have seen histograms which had more total entries than the corresponding table rows. However, the numbers were only slightly off. We resolved it to float inaccuracies (60e9 tuples in SF 10 lineitem cannot be stored without loss).


Don't know if that is the reason for your problem, but it seems to me that `estimate_null_value_ratio_of_column(...)` in `cardinality_estimator.cpp` might be vulnerable to the histogram size bug:
https://github.com/hyrise/hyrise/blob/6e429c14c4d71997227554bd020500fdf04581b5/src/lib/statistics/cardinality_estimator.cpp#L62

If the histogram size is bigger than the table size, the fraction in L. 62 evaluates to something > 1, and thus the total expression to something < 0
